### PR TITLE
Closes #3526: Update selected tab id after selected tab is removed.

### LINKED ISSUE #3526: browser-state: Update selected session if selected tab is removed
For `browser-session` parity.

### PR DESCRIPTION
**Pull request by pocmo**
Follow-up #3567

### CODE REVIEW

File: components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt [hunk 1]
```diff
@@ -26,16 +27,93 @@ internal object TabListReducer {
...
+            }
 
             is TabListAction.RemoveTabAction -> {
                 val tab = state.findTab(action.tabId)
+
                 if (tab == null) {
                     state
                 } else {
-                    state.copy(tabs = state.tabs - tab)
+                    val previousIndex = state.tabs.indexOf(tab)
```
Comment (NotWoods) [inline components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt]: nit: Move this computation inside the if then branch
```kt
val updatedSelection = if (state.selectedTabId == tab.id) {
    val previousIndex = state.tabs.indexOf(tab)
    findNewSelectedTabId(updatedTabList, tab.content.private, previousIndex)
} else ...
```

File: components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt [hunk 2]
```diff
@@ -26,16 +27,93 @@ internal object TabListReducer {
...
+    tabs: List<TabSessionState>,
+    isPrivate: Boolean,
+    previousIndex: Int
+): String? {
+    if (tabs.isEmpty()) {
+        // There's no tab left to select.
+        return null
+    }
+
+    val predicate: (TabSessionState) -> Boolean = { tab -> tab.content.private == isPrivate }
```
Comment (NotWoods) [inline components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt]: Can you omit the type if you do this?
```suggestion
    val predicate = { tab: TabSessionState -> tab.content.private == isPrivate }
```

File: components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt [hunk 3]
```diff
@@ -26,16 +27,93 @@ internal object TabListReducer {
...
+    val maxSteps = max(tabs.lastIndex - index, index)
+    if (maxSteps < 0) {
+        return null
+    }
+
+    // Try tabs oscillating near the index.
+    for (steps in 1..maxSteps) {
+        listOf(index - steps, index + steps).forEach { current ->
+            if (current >= 0 &&
+                current <= tabs.lastIndex &&
```
Comment (NotWoods) [inline components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt]: nit: `current in 0..tabs.lastIndex`

File: components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt (hunk 2 above)
Comment (pocmo) [reply components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt]: Yeah, that works. But do you think it's more readable? If I omit the type then Android Studio will show it itself and there's actually more text on the screen after the change :D

Comment (NotWoods) [reply components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt]: I didn't realize Android Studio does that, feel free to keep it as you originally wrote it.

**Pull request merged**